### PR TITLE
Populate prebuild and postbuild from procfile

### PR DESCRIPTION
**Pull request by tulikac**
This PR is to read the app.yaml (procfile) and populate the prebuild and postbuild commands.

The app.yaml has a structure like this

version: 1

pre-build: apt-get install xyz

build: pip install requirements.txt
  
post-build: |
  python manage.py makemigrations
  python manage.py migrate
  
run: gunicorn myapp.app --workers 5 --foo bar

If the app.yaml exists, I am overwriting any prebuild and postbuild appsettings that the customer might have.

### CODE REVIEW

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs
Comment (gpcastro) [inline src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs:26]: Please move to a separate file

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs
Comment (gpcastro) [inline src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs:24]: typo

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs
Comment (gpcastro) [inline src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs:345]: how about abstracting the section dealing with procfile parsing and error handling to a separate class altogether?

**Comment by gpcastro** [comment]
I don't see any tests for this change, and neither documentation on the use of procfiles. Also, is procfiles the actual name we'll be referring to this?

**Review by gpcastro — CHANGES_REQUESTED** [review]

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs
Comment (william-msft) [inline src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs:18]: @tulikac please merge with `main` and add integration tests

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs
Comment (tulikac) [reply src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs:26]: Done

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs
Comment (tulikac) [reply src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs:24]: Fixed

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs
Comment (tulikac) [reply src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs:345]: I have moved the parsing of the Procfile to a different class. However, it doesn't seem worth it to move the error handling as I would need to pass the logger and writers to the different class and they exist in the context of this class here.

**Comment by william-msft** [comment]
@tulikac please make sure the following tests also pass locally (currently failing in the pipeline): https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5597170&view=logs&j=e1391e23-97cc-5b71-5818-643679e655ef&t=10188d5a-2cf9-5c06-9eb2-4b00602c03d8&l=25

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs
Comment (tulikac) [reply src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs:18]: Added some integration tests

**Comment by william-msft** [comment]
@tulikac may you merge with `main` branch please. The following pipeline errors:
https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5692283&view=logs&j=3b2a2d55-d8d9-5a78-6715-6219efc4cdf5&t=81bad235-b840-5e51-fea1-46bc759086dc

Should be resolved with merge

File: tests/Oryx.Integration.Tests/BuildConfigurationFile/BuildConfigurationReadYaml.cs
Comment (william-msft) [inline tests/Oryx.Integration.Tests/BuildConfigurationFile/BuildConfigurationReadYaml.cs:20]: nit: please remove extra lines

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs
Comment (william-msft) [inline src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs:321]: please add a logging statement stating the prebuild and postbuild commands will be overwritten

File: tests/Oryx.Integration.Tests/BuildConfigurationFile/BuildConfigurationReadYaml.cs
Comment (william-msft) [inline tests/Oryx.Integration.Tests/BuildConfigurationFile/BuildConfigurationReadYaml.cs:263]: Please remove all extra new lines from all files

File: tests/Oryx.Integration.Tests/BuildConfigurationFile/BuildConfigurationReadYaml.cs
Comment (william-msft) [inline tests/Oryx.Integration.Tests/BuildConfigurationFile/BuildConfigurationReadYaml.cs:165]: nit: please correct comment indentation

File: tests/Oryx.Integration.Tests/BuildConfigurationFile/BuildConfigurationReadYaml.cs
Comment (tulikac) [reply tests/Oryx.Integration.Tests/BuildConfigurationFile/BuildConfigurationReadYaml.cs:20]: Done

File: tests/Oryx.Integration.Tests/BuildConfigurationFile/BuildConfigurationReadYaml.cs
Comment (tulikac) [reply tests/Oryx.Integration.Tests/BuildConfigurationFile/BuildConfigurationReadYaml.cs:263]: Done

File: tests/Oryx.Integration.Tests/BuildConfigurationFile/BuildConfigurationReadYaml.cs
Comment (tulikac) [reply tests/Oryx.Integration.Tests/BuildConfigurationFile/BuildConfigurationReadYaml.cs:165]: Remove this part of the documentation

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs
Comment (tulikac) [reply src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs:321]: Done

**Comment by qianz2** [comment]
@tulikac Can you rebase on the main branch please, I'm seeing some pipeline tests failures which can be resolved by doing that. Thanks!

**Review by william-msft — APPROVED** [review]

**Comment by william-msft** [comment]
@tulikac I brought you branch up to date with `main` for you :) I see `Oryx-validation` pipeline is pending. Once the pipeline passes, the merge button should be clickable.

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs
Comment (william-msft) [inline src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs:320]: @tulikac may you take a look at the failing tests:
https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5715683&view=logs&j=e1391e23-97cc-5b71-5818-643679e655ef&t=10188d5a-2cf9-5c06-9eb2-4b00602c03d8&l=19

These changes are failing tests for `DefaultBuildScriptGenerator`

**Pull request merged**
